tensorpipe/common/cuda_host_allocator.cc

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/common/cuda_host_allocator.h>
+
+#include <cuda_runtime.h>
+
+#include <tensorpipe/common/cuda.h>
+
+namespace tensorpipe {
+
+CudaHostAllocator::CudaHostAllocator(size_t numChunks, size_t chunkSize)
+    : numChunks_(numChunks),
+      chunkSize_(chunkSize),
+      data_(allocPinnedBuffer(numChunks * chunkSize)),
+      chunkAvailable_(numChunks, true) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  thread_ = std::thread([this, lock{std::move(lock)}]() mutable {
+    setThreadName("TP_CUDA_host_allocator_loop");
+    allocLoop(std::move(lock));
+  });
+}
+
+CudaHostAllocator::~CudaHostAllocator() {
+  join();
+}
+
+void CudaHostAllocator::alloc(size_t size, TAllocCallback callback) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (closed_) {
+    callback(TP_CREATE_ERROR(CudaHostAllocatorClosedError), THostPtr());
+    return;
+  }
+  TP_DCHECK(size <= chunkSize_);
+  pendingAllocations_.push_back(std::move(callback));
+  cv_.notify_all();
+}
+
+size_t CudaHostAllocator::getChunkLength() const {
+  return chunkSize_;
+}
+
+void CudaHostAllocator::close() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (closed_) {
+    return;
+  }
+  closed_ = true;
+  cv_.notify_all();
+}
+
+void CudaHostAllocator::join() {
+  close();
+  if (!joined_.exchange(true)) {
+    thread_.join();
+  }
+}
+
+void CudaHostAllocator::allocLoop(std::unique_lock<std::mutex> lock) {
+  for (;;) {
+    if (closed_ && pendingAllocations_.empty()) {
+      break;
+    } else {
+      cv_.wait(lock);
```

Comment:
This wake-up can happen in three cases: new request from the user, a previous chunk was returned to the pool, and the allocator was closed. The last case (i.e., closed) seems it could deserve a dedicated check after this wakeup?